teuthology/openstack.py

```diff
@@ -0,0 +1,375 @@
+        argv.append('/home/ubuntu/teuthology/teuthology/test/integration/openstack.yaml')
+        command = (
+            "source ~/.bashrc_teuthology ; " + self.teuthology_suite + " " +
+            " --machine-type openstack " +
+            " ".join(map(lambda x: "'" + x + "'", argv))
+        )
+        print self.ssh(command)
+
+    def setup(self):
+        if not self.cluster_exists():
+            self.create_security_group()
+            self.create_cluster()
+        return self.get_floating_ip(self.args.name)
+
+    def setup_logs(self):
+        loglevel = logging.INFO
+        if self.args.verbose:
+            loglevel = logging.DEBUG
+        teuthology.log.setLevel(loglevel)
+
+    def ssh(self, command):
+        client_args = {
+            'username': self.username,
+        }
+        if self.key_filename:
+            log.debug("using key " + self.key_filename)
+            client_args['key_filename'] = self.key_filename
+        ip = self.get_floating_ip(self.args.name)
+        log.debug("ssh " + self.username + "@" + str(ip) + " " + command)
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(
+            paramiko.AutoAddPolicy())
+        client.connect(ip, **client_args)
+        stdin, stdout, stderr = client.exec_command(command)
+        stdout.channel.settimeout(300)
+        out = ''
+        try:
+            out = stdout.read()
+            log.debug('teardown stdout ' + command + ' ' + out)
+        except Exception:
+            log.exception('teardown ' + command + ' failed')
+        err = stderr.read()
+        log.debug('teardown stderr ' + command + ' ' + err)
+        return out + ' ' + err
+
+    def verify_openstack(self):
+        try:
+            misc.sh("openstack server list")
+        except subprocess.CalledProcessError:
+            log.exception("openstack server list")
+            raise Exception("verify openrc.sh has been sourced")
+        if 'OS_AUTH_URL' not in os.environ:
+            raise Exception('no OS_AUTH_URL environment varialbe')
+        providers = (('cloud.ovh.net', 'ovh'),
+                     ('control.os1.phx2', 'redhat'),
+                     ('entercloudsuite.com', 'entercloudsuite'))
+        self.provider = None
+        for (pattern, provider) in providers:
+            if pattern in os.environ['OS_AUTH_URL']:
+                self.provider = provider
+                break
+        if not self.provider:
+            raise Exception('OS_AUTH_URL=' + os.environ['OS_AUTH_URL'],
+                            ' does is not a known OpenStack provider ' +
+                            str(providers))
+
+
+    def image(self):
+        if self.provider == 'ovh':
+            return 'Ubuntu 14.04'
+        elif self.provider == 'redhat':
+            return 'ubuntu-server-14.04-x86_64'
+        elif self.provider == 'entercloudsuite':
+            return 'GNU/Linux Ubuntu Server 14.04 Trusty Tahr x64'
```

Comment:
What's the rationale for hardcoding all these distros? And, why specifically Ubuntu Trusty?